c/tskit/tables.h

```diff
@@ -1425,6 +1468,36 @@ and is not checked for compatibility with any existing schema on this table.
 int tsk_node_table_extend(tsk_node_table_t *self, const tsk_node_table_t *other,
     tsk_size_t num_rows, const tsk_id_t *row_indexes, tsk_flags_t options);
 
+/**
+@brief Subset this table by keeping rows according to a boolean mask.
+
+@rst
+Deletes rows from this table and optionally return the mapping from IDs in
+the current table to the updated table. Rows are kept or deleted according to
+the specified boolean array ``keep`` such that for each row ``j`` if
+``keep[j]`` is false (zero) the row is deleted, and otherwise the row is
+retained. Thus, ``keep`` must be an array of at least ``num_rows``
+:c:type:`bool` values.
+
+If the ``id_map`` argument is non-null, this array will be updated to represent
+the mapping between IDs before and after row deletion. For row ``j``,
+``id_map[j]`` will contain the new ID for row ``j`` if it is retained, or
+:c:macro:`TSK_NULL` if the row has been removed. Thus, ``id_map`` must be an
+array of at least ``num_rows`` :c:type:`tsk_id_t` values.
+@endrst
+
+@param self A pointer to a tsk_node_table_t object.
+@param keep Array of boolean flags describing whether a particular
+    row should be kept or not. Must be at least ``num_rows`` long.
+@param options Bitwise option flags. Currently unused; should be
+    set to zero to ensure compatibility with later versions of tskit.
+@param id_map An array in which to store the mapping between new
+    and old IDs. If NULL, this will be ignored.
+@return Return 0 on success or a negative value on failure.
+*/
+int tsk_node_table_keep_rows(
+    tsk_node_table_t *self, const bool *keep, tsk_flags_t options, tsk_id_t *id_map);
```

Comment:
Overall, the API makes sense.  There's one technical issue that should be documented:

* When interacting with C functions, C++ code will tend to store things in a vector and obtain the pointer via the data member function.  For example, `x.data()` yields a pointer to the first element, which is then passed to a C API.
* C++ `vector<bool>` totally breaks all of what I just said.  Due to a grave lapse of judgement, whether bool is stored or individual elements are bit fields is implementation defined.

There's nothing we can do except to note that "the pointer from a c++ vector<bool> is not the right thing to pass in here".

Reply:
Or tell people to use `std::vector<uint8_t>` instead of `std::vector<bool>`, right?  As I understand it, that avoids the problem.

Reply:
Yes, so long as uint8_t  and bool have the same sizeof everywhere.  That's true for all platforms that I expect to encounter, but I'd still use a `static_assert` to check it at compile time.

Reply:
Yet another reason why I'm drifting away from C++.  From cppreference:

```
bool - type, capable of holding one of the two values: [true](https://en.cppreference.com/w/cpp/language/bool_literal) or [false](https://en.cppreference.com/w/cpp/language/bool_literal). The value of sizeof(bool) is implementation defined and might differ from 1.
```

And the first passes but the second fails:

```cpp
#include <cstdint>
#include <type_traits>

int
main(int, char**)
{
    static_assert(sizeof(std::uint8_t) == sizeof(bool), "sizeof failure");
    static_assert(std::is_same<std::uint8_t, bool>::value,
                  "uint8_t and bool aren't the same");
}

```

I think we have to say "be careful here" when using C++.  Too much is platform-dependent.

Reply:
Technically, true that they could be different – but as you said, on all platforms that we expect to encounter, they are in fact the same.  A matter of taste how much one wants to worry about that, I guess.  :->